### PR TITLE
Add Torch 2.4 Tests

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Add Torch 2.4 Tests. Includes
- Updating error messages to latest versions in PyTorch 2.4
- Zero grad must be now called at end of fit w new ckpting API for stateless optimizers, https://github.com/pytorch/pytorch/issues/133415

### CODE REVIEW

File: .github/workflows/pr-gpu.yaml [hunk 1]
```diff
@@ -39,8 +39,8 @@ jobs:
     strategy:
       matrix:
         include:
-        - name: gpu-3.11-2.3-2
-          container: mosaicml/pytorch:2.3.1_cu121-python3.11-ubuntu20.04
+        - name: gpu-3.11-2.342
```
Comment (dakinggg) [inline .github/workflows/pr-gpu.yaml:42]: typo here?

File: docs/source/doctest_fixtures.py [hunk 2]
```diff
@@ -117,7 +124,7 @@
 
 model = SimpleModel(num_channels, num_classes)
 
-optimizer = torch.optim.SGD(model.parameters(), lr=0.001)
+optimizer = DecoupledSGDW(model.parameters(), lr=0.001)
```
Comment (dakinggg) [inline docs/source/doctest_fixtures.py:127]: why this change?

**Review by dakinggg — APPROVED** [review]

File: docs/source/doctest_fixtures.py (hunk 2 above)
Comment (mvpatel2000) [reply docs/source/doctest_fixtures.py:127]: Due to https://github.com/pytorch/pytorch/issues/133415. Stateless optimizer causes problems

Comment (dakinggg) [reply docs/source/doctest_fixtures.py:127]: does that imply composer is broken with torch.optim.SGD?

Comment (mvpatel2000) [reply docs/source/doctest_fixtures.py:127]: Currently, yes 😬. We have workaround though (decoupledSGDW) and hoping to get into torch2.4.1 for a fix
